addons/website_sale/models/product.py

```diff
@@ -96,15 +96,18 @@ def _is_available_on_website(self, website_id):
         Change in this method should be reflected in `_get_website_pricelists_domain`.
         """
         self.ensure_one()
-        return self.website_id.id == website_id or (not self.website_id and (self.selectable or self.sudo().code))
+        company_id = self.env["website"].browse(website_id).company_id.id
+        return self.website_id.id == website_id or (not self.website_id and self.company_id.id in [False, company_id] and (self.selectable or self.sudo().code))
 
     def _get_website_pricelists_domain(self, website_id):
         ''' Check above `_is_available_on_website` for explanation.
         Change in this method should be reflected in `_is_available_on_website`.
         '''
+        company_id = self.env["website"].browse(website_id).company_id.id
         return [
             '|', ('website_id', '=', website_id),
             '&', ('website_id', '=', False),
+            '&', ('company_id', 'in', [False, company_id]),
```

Comment:
Same here about the company_id check not being checked if there is a website_id.
Add company_id check on top level

```python
return [
        '&', ('company_id', 'in', [False, company_id]),
        '|', ('website_id', '=', website_id),
        '&', ('website_id', '=', False),
        '|', ('selectable', '=', True), ('code', '!=', False),
    ]
```